File: reDiscover/CAKeyframeAnimation+Parametric.m
//
//  CAKeyframeAnimation+Parametric.m
//  reDiscover
//
//  Created by Teo Sartori on 02/04/14.
//  Copyright (c) 2014 Teo Sartori. All rights reserved.
//

#import "CAKeyframeAnimation+Parametric.h"

@implementation CAKeyframeAnimation (Parametric)

+ (id)animationWithKeyPath:(NSString *)path
                  function:(KeyframeParametricBlock)block
                 fromValue:(double)fromValue
                   toValue:(double)toValue {
    
    // get a keyframe animation to set up
    CAKeyframeAnimation *animation = [CAKeyframeAnimation animationWithKeyPath:path];
    
    // break the time into steps
    //  (the more steps, the smoother the animation)
    NSUInteger steps = 100;
    NSMutableArray *values = [NSMutableArray arrayWithCapacity:steps];
    double time = 0.0;
    double timeStep = 1.0 / (double)(steps - 1);
    
    for(NSUInteger i = 0; i < steps; i++) {
        double value = fromValue + (block(time) * (toValue - fromValue));
        [values addObject:[NSNumber numberWithDouble:value]];
        time += timeStep;
    }
    // we want linear animation between keyframes, with equal time steps
    animation.calculationMode = kCAAnimationLinear;
    // set keyframes and we're done
    [animation setValues:values];
    return(animation);
}

@end

File: CAKeyframeAnimation+Parametric.swift
//
//  CAKeyframeAnimation+Parametric.swift
//  reDiscover
//
//  Created by Teo on 21/08/15.
//  Copyright © 2015 Teo Sartori. All rights reserved.
//

import Foundation

typealias KeyframeParametricBlock = (time: Double) -> Double

extension CAKeyframeAnimation {
    
    /// Directly translated from my objc version CAKeyframeAnimation+Parametric.m/h so
    /// check that if something doesn't work.
    
    class func animation(keyPath path: String, function block: KeyframeParametricBlock,
        fromValue fromVal: Double, toValue toVal: Double) -> AnyObject {
        
        /// Get a keyframe animation set up.
        let animation = CAKeyframeAnimation(keyPath: path)
            
        /// Break the time into steps. The more stes the smoother the anim.
        let steps    = 100
        var time     = 0.0
        let timeStep = 1.0 / Double(steps-1)
        var values   = [Double]()
            
        for _ in 0 ..< steps {
            let value = fromVal + (block(time: time) * (toVal - fromVal))
            values.append(value)
            time += timeStep
        }
        
        animation.calculationMode   = kCAAnimationLinear
        animation.values            = values
            
        return animation
    }
    
}